### DISCUSSION TITLE
How does nelua's cache compilation works

### DISCUSSION THREAD

**Discussion by ghost** [edubart/nelua-lang#176] [Q&A] [answered]
How does nelua's cache compilation works

**Comment by edubart** [comment, marked as answer]
When you run `nelua main.nelua`, the compiler saves in the user cache directory (usually located at `$HOME/.cache/nelua/`) the following files:
* `$HOME/.cache/nelua/main.c` file, this is the representation of the whole program Nelua  in C source
* `$HOME/.cache/nelua/main` file (with `.exe` suffix on Windows), this is the binary of the whole program compiled to native machine executable.

When you compile the Nelua source code the current `$HOME/.cache/nelua/main.c` is read and compared with a new generated C file, in case both are the same (due to no source code changes) then the Nelua will skip invoking the C compiler, and simply run `$HOME/.cache/nelua/main` directly in case it exists. Otherwise, in case the newly generated C code does not match the old one or the binary doesn't exist, then a C compiler is invoked to compile the C code into a new binary, subsequently the binary is run. Using `--no-cache` options instructs the compiler to not read any files from the cache directory and to always generate the C code and invoke the C compiler, even if no source code changed, yet these files will still be generated in the cache directly, because they are temporary compilation files.

All this means the only purpose of the current cache system is to skip C recompilation between runs in case no source code changed. Yet for all runs the compiler analyze the entire Nelua source code being run, maybe in the future this may be improved with some kind of incremental compilation system to speed up compilation of large Nelua code bases.
